docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,228 @@
+---
+title: agent-upgrade
+authors:
+  - "@jhernand"
+creation-date: 2022-06-13
+last-updated: 2022-06-13
+---
+
+# Agent upgrade
+
+## Summary
+
+Usually the time since the agent is started and the cluster is installed is
+short. But in some situations it may be long, even many days, and during that
+long time the service may be upgraded and may require a newer version of the
+agent. There is currently no mechanism to upgrade the agent, the service
+silently works with the older agent assuming that nothing will go wrong. The
+only way to upgrade the agent would be to reboot the host, and we don't ask
+users to do that. Even if we did, it would be inconvenient, specially if there
+are many hosts or if rebooting a host requires manual intervention. This is a
+proposal to implement a mechanism to upgrade the agent without having to reboot
+the host.
+
+Note that all new agent versions require a rollout of the service pods (as the
+agent image name is delivered as an environment variable to the service). Not
+all rollouts of service pods come with a newer agent version.
+
+## Motivation
+
+Reduce compatibility concerns when designing and developing new features.
+
+Reduce the need to reboot hosts when there are changes in the service that
+require new versions of the agent.
+
+### Goals
+
+The agent image will be automatically upgraded to the version expected by the
+service before proceeding to the installation of the cluster.
+
+Note that in this context _version_ is really a container image name, like
+`registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141`.
+
+### Non-Goals
+
+The upgrade feature will not work for changes in the part of the agent that
+performs the initial registration since the upgrade part relies on the
+registration mechanism.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a developer I want the agent to be automatically upgraded to what the service
+expects, so that I can assume that the service will never proceed to install a
+cluster with a host that is running an incompatible version of the agent.
+
+### Implementation Details/Notes/Constraints
+
+A new `compatible-agent` validation will be added to the service to check if the
+agent is compatible.
+
+The logic for that validation will be a simple string comparison of the value
```

Comment:
According to my tests the latest version of the agent sends in the `discovery_agent_version` header only the tag of the image reference. Are we open to change that to the full image reference? You mention it has been changed at least twice in the past? What was the reason to change it to only the tag?

I am updating the document with your suggestion.

Reply:
That's exactly the problem :( we don't really know why it was changed, and then changed back (after a fix for the first change, so the second change broke the fix). Take a look at that PR's Jira ticket.

Reply:
Don't we just want to comapre the tag (sha256) anyway?
I don't think the registry or image name matters.
Moreover in case of disconnected installation, the agent fqdn (registry+name+tag) can pint to the internet and relay on the ICSP config on the node to resolve the image from the mirror registry.
So I think that we are better off if we just compare the tag.

Reply:
To say the truth i think better to compare full string. Not sure why we should match only tag, it doesn't save anything for us

Reply:
@tsorya what version the agent sends when installing in a disconnected env?
If we configure the service with the agent image from the mirror and that is what the agent is sending - great.
But if we rely on the ICSP to translate the agent fqdn in the service to the image from the mirror the agent might report something else.

Reply:
@eranco74 the agent extracts the tag from the image reference that it received from the service during host registration, and that is what it sends. Note that this is a pure string manipulation, no inspection of the image is involved.